[@types/yup] Optional subfields get marked as required with InferType

If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/yup` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
* @dhardtke
* @vtserman
* @MoretonBayRC
* @sseppola
* @YashdalfTheGray
* @vincentjames501
* @robertbullen
* @sat0yu
* @deskoh
* @mauricedb
* @kalley
* @elias-garcia
* @iansan5653

If you do not mention the authors the issue will be ignored.

**Describe the bug**

When when an `object` schema is created, it's Typescript types are faithfully inferred via `InferType`, but when that same schema is included as a nested schema within another object, the resulting inferred types incorrectly drop the optional status of any nested fields.

**To Reproduce**

Minimal repro here:
https://codesandbox.io/s/yup-infertype-nested-optional-issue-v63dd
```typescript
import { object, string, InferType } from "yup";

const nameSchema = object({
  first: string().notRequired(),
  last: string().notRequired()
});

type NameType = InferType<typeof nameSchema>;

const userSchema = object({
  name: nameSchema.notRequired()
});

type UserType = InferType<typeof userSchema>;

class User implements UserType {
  name?: NameType;
}
```

Results in error:
```
Property 'name' in type 'User' is not assignable to the same property in base type 'Id<Partial<Pick<{ name: { first: ...; last: ...; }; }, "name">> & Pick<{ name: { first: ...; last: ...; }; }, never>>'.
  Type 'Id<Partial<Pick<{ first: string; last: string; }, "first" | "last">> & Pick<{ first: string; last: string; }, never>>' is not assignable to type '{ first: string; last: string; }'.
    Property 'first' is optional in type 'Id<Partial<Pick<{ first: string; last: string; }, "first" | "last">> & Pick<{ first: string; last: string; }, never>>' but required in type '{ first: string; last: string; }'.ts(2416)
```

Based on the hints, `NameType` by itself is correct:
```
type NameType = {
    first?: string;
    last?: string;
}
```

but once included in UserType, the optionals get dropped:
```
type UserType = {
    name?: {
        first: string;
        last: string;
    };
}
```

**Expected behavior**
Optional state of fields is preserved for nested schemas

Hmm, I'm not sure that's it. I thought maybe the flags were not getting set correctly since my setup uses inherited tsconfigs but I double-checked the output and it breaks even with the flags set. Granted this is my actual code, not the minimal repro. I'll also try that

```
$ npx tsc --showConfig
{
    "compilerOptions": {
        "allowJs": true,
        "allowSyntheticDefaultImports": true,
        "declaration": false,
        "esModuleInterop": true,
        "forceConsistentCasingInFileNames": true,
        "isolatedModules": true,
        "lib": [
            "dom",
            "dom.iterable",
            "esnext"
        ],
        "module": "esnext",
        "moduleResolution": "node",
        "noEmit": true,
        "preserveSymlinks": true,
        "resolveJsonModule": true,
        "skipLibCheck": true,
        "strict": true,
        "strictNullChecks": true,
        "baseUrl": "./",
        "paths": {
            "~/*": [
                "src/*"
            ]
        },
        "jsx": "react",
        "target": "es5"
    },
    [...]

$ npx tsc
src/models/User.ts:14:3 - error TS2416: Property 'address' in type 'User' is not assignable to the same property in base type 'Id<Partial<Pick<{ id: number | undefined; } & { avatar: ({ id: number | undefined; } & { contentType: ...; contentSubType: ...; name: ...; fileType: ...; fileExtension: ...; fileContent: ...; type: ...; }) | null | undefined; ... 11 more ...; customDate1: number | ... 1 more ... | undefined; }, "id" | ... 12 more .....'.
  Type 'Id<Partial<Pick<{ address1: string | null | undefined; address2: string | null | undefined; city: string | null | undefined; state: string | null | undefined; zip: string | null | undefined; }, "address1" | ... 3 more ... | "zip">> & Pick<...>> | undefined' is not assignable to type '{ address1: string | null | undefined; address2: string | null | undefined; city: string | null | undefined; state: string | null | undefined; zip: string | null | undefined; } | undefined'.
    Type 'Id<Partial<Pick<{ address1: string | null | undefined; address2: string | null | undefined; city: string | null | undefined; state: string | null | undefined; zip: string | null | undefined; }, "address1" | ... 3 more ... | "zip">> & Pick<...>>' is not assignable to type '{ address1: string | null | undefined; address2: string | null | undefined; city: string | null | undefined; state: string | null | undefined; zip: string | null | undefined; }'.
      Property 'address1' is optional in type 'Id<Partial<Pick<{ address1: string | null | undefined; address2: string | null | undefined; city: string | null | undefined; state: string | null | undefined; zip: string | null | undefined; }, "address1" | ... 3 more ... | "zip">> & Pick<...>>' but required in type '{ address1: string | null | undefined; address2: string | null | undefined; city: string | null | undefined; state: string | null | undefined; zip: string | null | undefined; }'.

14   address?: Address;
     ~~~~~~~


Found 1 error.
```

Yeah the minimal repro works for me too even with the correct `tsconfig.json`

```
~/typesyupbug $ npx tsc --showConfig
{
    "compilerOptions": {
        "strict": true,
        "strictNullChecks": true
    },
    "files": [
        "./test.ts"
    ]
}
~/typesyupbug $ npx tsc test.ts
test.ts:10:3 - error TS2741: Property 'a' is missing in type '{}' but required in type '{ a: number; }'.

10   b: {
     ~

  test.ts:5:7
    5       a: number()
            ~~~~~~~~~~~
    'a' is declared here.
  test.ts:4:3
    4   b: object({
        ~~~~~~~~~~~
    5       a: number()
      ~~~~~~~~~~~~~~~~~
    6   })
      ~~~~
    The expected type comes from property 'b' which is declared here on type 'Id<Partial<Pick<{ b: { a: ...; }; }, "b">> & Pick<{ b: { a: ...; }; }, never>>'


Found 1 error.

~/typesyupbug $ npx tsc --version
Version 3.9.3
```

--- a/types/yup/index.d.ts
+++ b/types/yup/index.d.ts
@@ -639,7 +639,9 @@ type KeyOfUndefined<T> = {
 type PreserveNull<T> = T extends null ? null : never;
 type PreserveUndefined<T> = T extends undefined ? undefined : never;
 type PreserveOptionals<T> = PreserveNull<T> | PreserveUndefined<T>;
-type Id<T> = { [K in keyof T]: T[K] };
+type Id<T> = {
+    [K in keyof T]: T[K] extends object ? InnerInferType<T[K]> : T[K];
+};
 type RequiredProps<T> = Pick<T, Exclude<keyof T, KeyOfUndefined<T>>>;
 type NotRequiredProps<T> = Partial<Pick<T, KeyOfUndefined<T>>>;
 type InnerInferType<T> =

--- a/types/yup/yup-tests.ts
+++ b/types/yup/yup-tests.ts
@@ -798,7 +798,15 @@ const personSchema = yup.object({
     // handle nested optional property
     address: yup.object({
         line1: yup.string().required(),
-        line2: yup.string().optional()
+        line2: yup.string().optional(),
+        area: yup.object({
+            country: yup.string().required(),
+            region: yup.string().optional()
+        }).required(),
+    }).required(),
+    addressBook: yup.object({
+        phoneNumbers: yup.array(yup.string()).defined(),
+        lastUpdated: yup.date().optional()
     }).required(),
     email: yup
         .string()
@@ -856,7 +864,11 @@ const minimalPerson: Person = {
     canBeNull: null,
     mustBeAString: '',
     address: {
-        line1: '123 Fake Street'
+        line1: '123 Fake Street',
+        area: { country: 'Australia' }
+    },
+    addressBook: {
+        phoneNumbers: ['123-456-789']
     }
 };
 
@@ -871,7 +883,11 @@ const person: Person = {
     children: null,
     address: {
         line1: 'Unit 1',
-        line2: '456 Fake Street'
+        line2: '456 Fake Street',
+        area: { country: 'America', region: 'WA' }
+    },
+    addressBook: {
+        phoneNumbers: []
     }
 };
 
@@ -886,6 +902,12 @@ person.friends = new Set(["Amy", "Beth"]);
 
 // $ExpectError
 person.address = {};
+// $ExpectError
+person.address = { area: {}, line1: '' };
+// $ExpectError
+person.addressBook = {};
+// $ExpectError
+person.addressBook = { phoneNumbers: null };
 // $ExpectError
 person.gender = 1;
 // $ExpectError

--- a/types/yup/yup-tests.ts
+++ b/types/yup/yup-tests.ts
@@ -795,6 +795,11 @@ enum Gender {
 const personSchema = yup.object({
     firstName: yup.string().defined(), // $ExpectType StringSchema<string>
     gender: yup.mixed<Gender>().defined().oneOf([Gender.Male, Gender.Female]),
+    // handle nested optional property
+    address: yup.object({
+        line1: yup.string().required(),
+        line2: yup.string().optional()
+    }).required(),
     email: yup
         .string()
         .nullable()
@@ -850,6 +855,9 @@ const minimalPerson: Person = {
     gender: Gender.Female,
     canBeNull: null,
     mustBeAString: '',
+    address: {
+        line1: '123 Fake Street'
+    }
 };
 
 const person: Person = {
@@ -861,6 +869,10 @@ const person: Person = {
     isAlive: null,
     mustBeAString: '',
     children: null,
+    address: {
+        line1: 'Unit 1',
+        line2: '456 Fake Street'
+    }
 };
 
 person.email = 'some@email.com';
@@ -872,6 +884,8 @@ person.children = ['1', '2', '3'];
 person.children = undefined;
 person.friends = new Set(["Amy", "Beth"]);
 
+// $ExpectError
+person.address = {};
 // $ExpectError
 person.gender = 1;
 // $ExpectError

